spec/src/main/asciidoc/Structure and format.adoc

```diff
@@ -0,0 +1,81 @@
+== Structure and format
+
+=== Terms
+
+.Terms and definitions
+Configuration:: An optional _top-level value_, with a collection of zero or more nested _configuration entries_ (organized by _configuration key_).
+  Also informally known as a "configuration tree node" or "configuration node".
+  A configuration with no value and no nested entries is considered to be an _empty configuration_.
+
+Configuration entry:: An _configuration key_ and _configuration value_ combination.
+
+Configuration key:: A non-empty string which uniquely identifies a configuration entry within a single nesting level of a configuration.
+  A configuration key *must not* contain embedded newlines or other control characters.
+  A configuration key string *must* be normalized using https://www.unicode.org/reports/tr15/#Norm_Forms)[Unicode Normalization Form C (NFC)].
+
+Configuration path:: A sequence of _configuration keys_ which uniquely identify a descendant _configuration entry_ within a configuration, at an arbitrary level of nesting.
+  Each configuration key within the sequence identifies one level of nested configuration.
+
+Configuration source:: A provider of configuration data in _raw value_ form. [TODO: high level structure of config sources, ordering etc]
+
+Configuration value:: An object which is stored within a _configuration_ and is accessible at the _configuration path_ which corresponds to that _configuration_.
+
+Empty configuration:: A configuration which has no _top-level value_ and zero nested _configuration entries_.
+
+Raw value:: A _configuration value_ in its uncoverted (string) form.
+
+Root:: The _configuration_ which is identified by a empty (zero-length) _configuration path_.
+  Every _configuration_ is its own _root_ when considered in isolation.
+
+Top-level value:: A _configuration value_ which resides at the _root_ of a _configuration_.
+
+== Standard mapping formats
+
+
+
+=== Mapping to `properties` format
+
+The `properties` file format corresponds to a text file in a format that is loadable by `java.util.Properties`.
+Note that `properties` files recognized by this specification *must* be encoded using the UTF-8 character encoding.
+Each key in the `properties` file is treated as a _configuration path_, with dot (`.`) characters acting as separators between the individual component _configuration keys_.
+
+==== Property key encoding
+
+Since an individual _configuration key_ may contain a dot (`.`) character, using a `.` character to separate `property` keys into a _configuration path_ poses a problem.
+Because of this, an encoding scheme should be used when dealing with keys that contain `.` characters.
+
+When properties are loaded from a `properties` file or from the system `properties` map, each encountered key is transformed into a _configuration path_ as follows:
```

Comment:
Note: I am assuming that the entire properties file must be loaded and processed in order to satisfy these transformations.

For this reason I think a hierarchical config source makes more sense than a "flat" config source for this use case. Which makes me wonder, what _is_ the use case for a "flat" config source? IMO we will never have a single uniform set of rules for composing a hierarchy from a flat source - it will *always* depend on the actual format.